modules/ipp/scanner.go

```diff
@@ -4,40 +4,73 @@ package ipp
 
 //TODO: Clean up these imports
 import (
-	//"bytes"
+	"bytes"
+	"crypto/sha256"
 	"encoding/binary"
-	//"errors"
+	"errors"
 	//"fmt"
-	//"io"
-	"net/http"
+	"io"
+	"io/ioutil"
+	"mime"
+	"net"
+	"net/url"
 	"strconv"
 	"strings"
-	//"net"
-	//"net/url"
-	//"time"
+	"time"
 
 	log "github.com/sirupsen/logrus"
 	"github.com/zmap/zgrab2"
+	"github.com/zmap/zgrab2/lib/http"
 )
 
 const (
-	ContentType string = "application/ipp"
+	ContentType         string = "application/ipp"
+	VersionsSupported   string = "ipp-versions-supported"
+	CupsVersion         string = "cups-version"
+	PrinterURISupported string = "printer-uri-supported"
+	NoDestinationsAdded string = "No destinations added."
 )
 
+var (
+	// ErrRedirLocalhost is returned when an HTTP redirect points to localhost,
+	// unless FollowLocalhostRedirects is set.
+	ErrRedirLocalhost = errors.New("Redirecting to localhost")
+
+	// ErrTooManyRedirects is returned when the number of HTTP redirects exceeds
+	// MaxRedirects.
+	ErrTooManyRedirects = errors.New("Too many redirects")
+
+	// TODO: Explain this error
+	ErrVersionNotSupported = errors.New("IPP version not supported")
+	Versions            = [...]version {{Major: 2, Minor: 1}, {Major: 2, Minor: 0}, {Major: 1, Minor: 1}, {Major: 1, Minor: 0},}
+)
+
+type scan struct {
+	connections []net.Conn
+	transport   *http.Transport
+	client      *http.Client
+	results     ScanResults
+	url         string
+}
+
 //TODO: Tag relevant results and exlain in comments
 // ScanResults instances are returned by the module's Scan function.
 type ScanResults struct {
 	//TODO: ?Include the request sent as well??
-	//TODO: Include a full response or at least a blob in the data (at least in verbose mode)
-	//Response *http.Response `json:"response,omitempty"`
+	Response     *http.Response `json:"response,omitempty" zgrab:"debug"`
+	CUPSResponse *http.Response `json:"cups_response,omitempty" zgrab:"debug"`
 
-	MajorVersion int8 `json:"version_major"`
-	MinorVersion int8 `json:"version_minor"`
+	// RedirectResponseChain is non-empty if the scanner follows a redirect.
```

Comment:
It looks like these new fields need to be added to https://github.com/zmap/zgrab2/blob/46029a8176d3a3d0e141fa889795d7c5678f3e37/zgrab2_schemas/zgrab2/ipp.py